api/agent/protocol/json.go

```
@@ -0,0 +1,133 @@
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		bb := new(bytes.Buffer)
+		_, err = bb.ReadFrom(req.Body)
+		if err != nil {
+			return err
+		}
+
+		err = stdin.Encode(bb.String())
+		if err != nil {
+			return err
+		}
+		_, err = io.WriteString(h.in, `",`)
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		defer bb.Reset()
+		defer req.Body.Close()
+	}
+	_, err = io.WriteString(h.in, `"headers:"`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	err = stdin.Encode(req.Header)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	_, err = io.WriteString(h.in, `"}`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	return nil
+}
+
+func (h *JSONProtocol) Dispatch(w io.Writer, req *http.Request) error {
+	err := h.DumpJSON(w, req)
+	if err != nil {
+		return respondWithError(
+			w, fmt.Errorf("unable to write JSON into STDIN: %s", err.Error()))
+	}
+	jout := new(JSONIO)
+	dec := json.NewDecoder(h.out)
+	if err := dec.Decode(jout); err != nil {
+		return respondWithError(
+			w, fmt.Errorf("unable to decode JSON response object: %s", err.Error()))
+	}
+	if rw, ok := w.(http.ResponseWriter); ok {
+		// this has to be done for pulling out:
+		// - status code
+		// - body
+		// - headers
+		for k, vs := range jout.Headers {
+			for _, v := range vs {
+				rw.Header().Add(k, v) // on top of any specified on the route
+			}
+		}
+		rw.WriteHeader(jout.StatusCode)
+		_, err = rw.Write([]byte(jout.Body)) // TODO timeout
+		if err != nil {
+			return respondWithError(
+				w, fmt.Errorf("unable to write JSON response object: %s", err.Error()))
+		}
+	} else {
+		// logs can just copy the full thing in there, headers and all.
+		err = json.NewEncoder(w).Encode(jout)
+		if err != nil {
+			return respondWithError(
+				w, fmt.Errorf("error writing function response: %s", err.Error()))
+		}
+	}
+	return nil
+}
+
+func respondWithError(w io.Writer, err error) error {
```

i think we should probably just return the error from `Dispatch` rather than writing it back to the user / logs from in here. the `http` format follows this convention, and I think now that we have multiple 'hot' formats going on it makes sense for us to follow that so that we can possibly format errors to the user in a way that will be friendly to them in one place (rather than having to basically make a different `respondWithError` for every format). so I think get rid of this `respondWithError` function and just return the error in `Dispatch` -- `Dispatch`'s caller is hooked up to handle receiving the error and returning it to the user, so we don't need to do it in `Dispatch` (and caller will return json object to user or hopefully in soon future `async` will add this error to the `Call` struct -- don't worry about doing that here)